File: app/src/main/java/martakonik/flashcards/flashcardsList/WordsListAdapter.java
package martakonik.flashcards.flashcardsList;


import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import java.util.List;

import martakonik.fiszki.domain.models.Flashcard;
import martakonik.flashcards.R;

public class WordsListAdapter extends RecyclerView.Adapter<WordsListAdapter.ViewHolder> {

    private List<Flashcard> flashcards;

    public WordsListAdapter(List<Flashcard> flashcards) {
        this.flashcards = flashcards;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View itemView = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.flashcard_list_row, parent, false);
        return new ViewHolder(itemView);
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        Flashcard flashcard = flashcards.get(position);
        holder.mTextViewWord.setText(flashcard.getWord());
        holder.mTextViewTranslation.setText(flashcard.getTranslation());
    }

    @Override
    public int getItemCount() {
        return flashcards.size();
    }

    public class ViewHolder extends RecyclerView.ViewHolder {
        TextView mTextViewWord = itemView.findViewById(R.id.text_view_translation);
        TextView mTextViewTranslation = itemView.findViewById(R.id.text_view_word);

        public ViewHolder(View itemView) {
            super(itemView);
        }
    }
}

File: app/src/main/java/martakonik/flashcards/flashcardsList/WordsListAdapter.kt
package martakonik.flashcards.flashcardsList


import android.content.res.Resources
import android.support.v4.app.FragmentManager
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.ViewGroup
import io.realm.OrderedRealmCollection
import io.realm.RealmRecyclerViewAdapter
import martakonik.flashcards.databinding.FlashcardListRowBinding
import martakonik.flashcards.models.Flashcard

class WordsListAdapter(
        data: OrderedRealmCollection<Flashcard>,
        private val supportFragmentManager: FragmentManager,
        private val resources: Resources
) : RealmRecyclerViewAdapter<Flashcard, WordsListAdapter.ViewHolder>(data, true) {

    init {
        setHasStableIds(true)
    }

    override fun getItemId(index: Int): Long = getItem(index)?.id?.toLong() ?: 0

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder =
            ViewHolder(FlashcardListRowBinding.inflate(LayoutInflater.from(parent.context), parent, false))

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.binding.apply {
            viewModel = FlashcardListViewModel(
                    getItem(position),
                    supportFragmentManager,
                    checkIfAddHeader(position),
                    resources
            )
            executePendingBindings()
        }
    }

    private fun checkIfAddHeader(position: Int): Boolean {
        when (position) {
            0 -> if (getItem(position)?.partOfBoxId == 0) return true
            itemCount - 1 -> return false
            else -> if (getItem(position - 1)?.partOfBoxId != getItem(position)?.partOfBoxId) {
                return true
            }
        }
        return false
    }

    inner class ViewHolder(val binding: FlashcardListRowBinding) : RecyclerView.ViewHolder(binding.root)
}
